### PR TITLE
Excel Reader Refactor - Base Class Introduction

### PR DESCRIPTION
**Pull request by WillAyd**
This is to provide further enablement of new Excel readers like openpyxl and potentially one for reading binary files.

The main goal here was the introduction of a base reader class which defines the properties / methods that any subclass needs to define to plug in to.

This ultimately could be refactored into a sub-directory which splits readers into it's own module

### CODE REVIEW

File: pandas/io/excel.py [hunk 1]
```diff
@@ -620,6 +532,120 @@ def _parse_cell(cell_contents, cell_typ):
...
+        # can't pass to get_filepath_or_buffer()
+        if _is_url(filepath_or_buffer):
+            filepath_or_buffer = _urlopen(filepath_or_buffer)
+        elif not isinstance(filepath_or_buffer, (ExcelFile, xlrd.Book)):
+            filepath_or_buffer, _, _, _ = get_filepath_or_buffer(
+                filepath_or_buffer)
+
+        if isinstance(filepath_or_buffer, xlrd.Book):
+            self.book = filepath_or_buffer
+        elif hasattr(filepath_or_buffer, "read"):
```
Comment (WillAyd) [inline pandas/io/excel.py]: Note that I changed this condition slightly to remove the `not isinstance(filepath_or_buffer, xlrd.Book)` as it was unnecessary given the preceding statement

**Review by jreback — COMMENTED** [review]
can you create another level here so we can split these out to other files, e.g.
pandas.io.excel
base
xlrd
styler
....

can do in a followup

File: pandas/io/excel.py [hunk 2]
```diff
@@ -375,60 +375,20 @@ def read_excel(io,
...
-            if hasattr(filepath_or_buffer, 'seek'):
-                try:
-                    # GH 19779
-                    filepath_or_buffer.seek(0)
-                except UnsupportedOperation:
-                    # HTTPResponse does not support seek()
-                    # GH 20434
-                    pass
+    def get_sheet_by_name(self, name):
+        raise NotImplementedError
```
Comment (jreback) [inline pandas/io/excel.py]: use AbstractMethodError

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment on this PR. So can merge this and split up.

File: pandas/io/excel.py (hunk 2 above)
Comment (gfyoung) [reply pandas/io/excel.py]: Is there a reason for not using `abc` for this?

**Comment by jreback** [comment]
lgtm @WillAyd

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**